pkg/asset/installconfig/gcp/services.go

```diff
@@ -0,0 +1,301 @@
+package gcp
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/cloudresourcemanager/v3"
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/dns/v1"
+	"google.golang.org/api/file/v1"
+	"google.golang.org/api/iam/v1"
+	"google.golang.org/api/option"
+	serviceusage "google.golang.org/api/serviceusage/v1beta1"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+type serviceNameGCP string
+type serviceVersionGCP string
+
+const (
+	computeServiceNameGCP   serviceNameGCP = "compute"
+	containerServiceNameGCP serviceNameGCP = "container"
+	storageServiceNameGCP   serviceNameGCP = "storage"
+
+	serviceVersionGCPv1 serviceVersionGCP = "v1"
+)
+
+// FormatGCPEndpointInput is the structure containing input variables for formatting the GCP Service Endpoints.
+type FormatGCPEndpointInput struct {
+	// SkipPath should be set to true when the path should not be added to the
+	// formatted endpoint. When the path is added, an example endpoint of
+	// https://compute-exampleendpoint.p.googleapis.com would be formatted as
+	// https://compute-exampleendpoint.p.googleapis.com/compute/v1/
+	SkipPath bool
+}
+
+// FormatGCPEndpointList will format the list of GCP Service Endpoints to match the expected url
+// for WithEndpoint or BasePath override endpoint options.
+func FormatGCPEndpointList(endpoints []configv1.GCPServiceEndpoint, input FormatGCPEndpointInput) ([]configv1.GCPServiceEndpoint, error) {
+	// The endpoints are modified to include the path
+	modifiedEndpoints := []configv1.GCPServiceEndpoint{}
+	for _, se := range endpoints {
+		formattedURL, err := FormatGCPEndpoint(se.Name, se.URL, input)
+		if err != nil {
+			return nil, fmt.Errorf("failed to format GCP service endpoint %s: %w", se.Name, err)
+		}
+
+		logrus.Debugf("Formatted GCP service endpoint %s: %s", se.Name, formattedURL)
+		modifiedEndpoints = append(modifiedEndpoints, configv1.GCPServiceEndpoint{Name: se.Name, URL: formattedURL})
+	}
+	return modifiedEndpoints, nil
+}
+
+// FormatGCPEndpoint will format the endpoint to ensure that the string is in the format that would be
+// accepted by both options (WithEndpoint and BasePath override).
+func FormatGCPEndpoint(service configv1.GCPServiceEndpointName, endpoint string, input FormatGCPEndpointInput) (string, error) {
```

Comment:
returning an error from this function, forces every caller to check the error and makes the code more verbose/less readable. Do we need to check for errors in this function? Shouldn't these errors have been caught by validation of the install config?